rslib/src/collection/backup.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::{
+    fs::{read_dir, remove_file, DirEntry, File},
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+use chrono::prelude::*;
+use itertools::Itertools;
+use zip::{write::FileOptions, CompressionMethod, ZipWriter};
+use zstd;
+
+use crate::prelude::*;
+
+const BACKUP_FORMAT_STRING: &str = "backup-%Y-%m-%d-%H.%M.%S.colpkg";
+/// Backups in the last [NO_THINNING_SECS] seconds are not thinned.
+const NO_THINNING_SECS: u64 = 2 * 24 * 60 * 60;
+/// At most one backup every [THINNING_INTERVAL_SECS] seconds is kept.
+const THINNING_INTERVAL_SECS: i64 = 24 * 60 * 60;
+
+pub fn backup(col_path: &str, out_dir: &str) -> Result<()> {
+    let out_file = File::create(out_path(out_dir))?;
+    let mut zip = ZipWriter::new(out_file);
+    let options = FileOptions::default().compression_method(CompressionMethod::Stored);
+
+    let col_file = File::open(col_path)?;
+    let encoded_col = zstd::encode_all(col_file, 0)?;
+
+    zip.start_file("collection.anki2", options)?;
+    zip.write_all(&encoded_col)?;
+    zip.start_file("media", options)?;
+    zip.write_all(b"{}")?;
+    zip.finish()?;
+
+    Ok(())
+}
+
+pub fn thin_backups(backup_dir: &str) -> Result<()> {
+    let backups = read_dir(backup_dir)?
+        .filter_map(|entry| entry.ok().and_then(Backup::from_entry))
+        .sorted_unstable_by_key(|entry| entry.timestamp)
+        .rev();
+    let mut last_kept_backup_time = TimestampSecs(i64::MAX);
+
+    for backup in backups {
+        let keep_this_backup = backup.timestamp.elapsed_secs() < NO_THINNING_SECS
+            || last_kept_backup_time.elapsed_secs_since(backup.timestamp) > THINNING_INTERVAL_SECS;
```

Comment:
Yes, I thought about that, but deliberately decided against using calendar days. It's advised to study before going to sleep, which is around midnight for a significant amount of users. That means that you could study every 'Anki day', but not every calendar day, because you might study on Monday 11 pm, Wednesday 1 am, and have the day boundary set to 4 am. If you study again at 11 pm on Wednesday, you end up with a 48h gap, because there is no backup for the calendrical Tuesday.

Of course, in this scenario, one of the Wednesday backups is discarded with my implementation as well. I'm not sure what the best solution is here. A few ideas:

1. Keep the current implementation but with 12h intervals. If we assume that people tend to study at around the same times of the day, there should rarely be gaps in the daily backups. On the other hand, at most two per day will be stored, which seems moderate.
2. Use Anki days. This may be a bit fragile if users change the rollover time. Also, some seem to have set it to an unreasonable time.
3. For every day, keep the backup closest to 12 am, even if it's across a day boundary.

Do you have other ideas? I think I'd prefer no. 3.

On a different note, should we encode the timezone in the file name? We can always fall back on the local one for older backups, of course.

P.s. Hope I didn't confuse am/pm anywhere. I'm not used to the 12-hour clock. 😅

Reply:
> For every day, keep the backup closest to 12 am, even if it's across a day boundary.

This feels a bit complicated to code, and to convey to the user. Couldn't we just group the backups into days, then keep the first or last backup of each ordered group, to achieve roughly similar results?

I probably wouldn't bother with a timezone - moves to different timezones are rare, these files are not intended to be shared with others, and since we only start thinning after 2 days, a timezone move shouldn't interrupt any immediate backups. If we add a TZ to the filename and start trying to read it back in when thinning/displaying the list, it adds a bit of complexity to the code, and adds some new failure cases like an invalid TZ name. WDYT?

Reply:
What exactly do you want to convey to the user? I thought we'd just offer a setting "backups for the last x days", and optionally a tooltip like "Anki keeps all backups from the last two days, and about 1 for older days." I can't imagine a user will mind if they look into the backup folder and there are two backups for the same date.
Wouldn't grouping result in the same gaps as when using ce days? How would it work in the above example? (Mon 11pm, Wed 1am, Wed 11pm.)

I just thought we better add the timezone now, if we possibly have use for it in the future, but as you point out, that probably won't be the case.

Reply:
>  How would it work in the above example? (Mon 11pm, Wed 1am, Wed 11pm.)

My thinking was that if we use the first backup of each day, they get Mon 11, and Wed 1am. If we use the last backup, they get Wed 11pm instead of Wed 1am.

I do see the problem you are trying to solve, and it's totally valid, I'm just worried that trying to me "smart" about how we handle this case might end up causing confusion for users. On mobile devices where space tends to be tight, backups for a 200MB collection will add up quickly, and I suspect some users may prefer that multiple ones aren't taken in a given day, as it reduces the total number of days they're willing to keep to stay under their space budget. Having said that though, there are probably other users who are more concerned with getting a backup in cases like the one you presented above. Hmm.

Could you give me some pseudo-code  for how you see approach 3 working? Maybe I have not understood what you are proposing fully.

And on the topic of mobile, mobile presents an added complication, as the user "closes" the app each time they switch away to a different one. We need to be able to take backups when they do it, but we don't want hundreds of backups a day, so we probably need some way to skip the backup if a minimum amount of time has not passed since the previous one. This could potentially be useful for the computer too, for people Anki devs and add-on authors, who may be opening and closing the app a lot :-) But there are cases like before a full sync where we want to ensure a backup is taken, regardless of how much time has elapsed.

> I just thought we better add the timezone now, if we possibly have use for it in the future, but as you point out, that probably won't be the case.

It's a reasonable point. My gut tells me YAGNI here, as it hasn't come up so far with the existing timestamp setup that has been in place for years. But I there have been times in the past where I've felt that way, and later come to regret it, so I can't say with certainty :wink:

Reply:
Wrote the pseudocode and realised that it doesn't even solve the problem in all cases. Maybe the solution is much simpler: We just use the last backup of each day, but on days we have no backup for, we take the first one of the following day instead. If we want, we can add an additional condition, like it must be from within the first x hours of the new day, or it mustn't be closer than x hours to the backup picked for the new day.

<details><summary>Pseudocode (feel free to ignore it)</summary>

```python
# first two days are handled separately
day_to_backup = DateTime.today().set_time(12, 0, 0).advance_by_days(-2)
last_backup = None
for backup in backups_in_newest_first_order:
    if backup.datetime.is_earlier_than(day_to_backup):
        # day_to_backup exceeded; ideal backup is this or the last one
        if backup.datetime.until(day_to_backup) < day_to_backup.until(last_backup.datetime):
            backup.keep_it = True
        else:
            last_backup.keep_it = True
        day_to_backup.advance_by_days(-1)
    last_backup = backup
```

</details>

> so we probably need some way to skip the backup if a minimum amount of time has not passed since the previous one.

Or do some conservative thinning to ensure the most recent version is backed up? Either way, it's tricky because the user may deliberately close the app to force a backup before doing something risky. If it goes wrong, they will not be amused if Anki was smart and skipped or deleted that one backup. 😃
Maybe that's more likely to happen on the desktop client though, where people do more complex operations, and know what counts as a "close".

Reply:
AnkiMobile provides a 'create backup now' option in the preferences/backup screen, which users can use if they want to ensure a backup is taken before they do something experimental.

The problem with unconditionally creating backups and then thinning them afterwards is that we're using the user's flash memory and CPU/battery each time - with a 200MB collection and switching to and from the app every 30 seconds to look words up in a dictionary as they're being created, that's a lot of wasted resources. And even with a throttle, the user is in a better place than they currently are with the computer version, where hours or days can go by without a backup if they don't close their profile. At one point we should probably think about adding timer-based backup creation as well. With the backing up happening in the backend, we could potentially do it while holding the mutex the entire time, preventing any issues with timers or GUI code that tries to access the DB while we close it and copy the data.

I think we need to strike a balance here between potential work lost, and the computing resources required to prevent it. Users have automatic saving and an undo system as a first line of defense, and having to restore from a backup because of a bug is rare - it's usually more about recovering from an accidental mistakes, where the user hasn't thought to make a manual backup first, and has already lost their undo history. Losing 30 minutes of work in that situation is frustrating, but isn't the end of the world. If you feel 30 minutes is too long, maybe we should be making it configurable.

If we're going to add in some smarts to try to cover the early morning study case, perhaps it's worth again considering providing a 'keep x weeks of backups' option as well, as after the first set of conditions, it should be pretty easy to add more. I'm conscious that users sometimes fail to detect problems like deleted cards for some time, and having longer term backups can occasionally save people. One possible approach we could take:

- we gather all available backups into a Vec<Backup>, where Backup has .day and .week ordinals, and a .keep bool
- we apply each of our limits in turn, eg:
- first setting keep=true on all backups within 48 hours (especially if we're throttling them, extra thinning is probably not necessary)
- then setting keep=true on the backups within a 2-7 days that we want to keep
- then setting keep=true on one backup per week
- finally, removing any items without the keep flag set

WDYT?

Reply:
What you say about mobile backups makes sense. I won't try to argue what the best value for the minimum interval between backups is.

> I'm conscious that users sometimes fail to detect problems like deleted cards for some time, and having longer term backups can occasionally save people.

Definitely. I didn't realise it at first, but of course you can restore a deleted deck from a backup and import it into your actual collection, without losing any progress made since. With that in mind, there might even be value in keeping extremely old backups, extremely thinned accordingly. However, as we want the user to decide how much backups to keep, that's not easily possible.
Maybe week-based thinning is the right compromise, I don't know. What setting(s) would you present to the user in this case? One of the reasons I went with days, was that it makes for an easy user decision: "How many days to keep?"
Directly asking for the number doesn't seem practical anymore. If you trigger a lot of backups on one day, that will dispose of the old ones. But once these new backups are old enough to get thinned, you end up with less backups than you asked for.

Reply:
Couldn't we just have "keep up to x daily backups" and "keep up to y weekly backups"? As long as x > 0 and the user has studied in the given week, there should always be at least one daily backup in the weekly period that can be used as the weekly backup. 5 might be a sensible default for x if we have a 48 hour exclusion zone, which should give ~1 week of daily backups. If we wanted to keep the options down, the daily setting could potentially be hidden, and just set to 5 when y>0, and backups disabled completely when y=0. Doesn't hurt much to keep it visible though, as it's sort of documenting what Anki is doing (and most people never tweak the backup options anyway)

Reply:
(adding a 'up to z monthly backups' wouldn't be hard either)